### PR TITLE
Add uiSchema explainer to readthedocs form file

### PR DESCRIPTION
**Pull request by CodeKrakken**

### CODE REVIEW

**Comment by CodeKrakken** [comment]
> This introduces the idea of subroutines as "gosubs", I don't think that's a common term though. If I read this, I would think that gosubs are the only way to have dynamic form field behaviour. But aren't they just one block of all the blocks? @CodeKrakken

@lukestanley I would have said the same but there's a page on gosubs here: https://kendraio-app.readthedocs.io/en/latest/workflow/blocks/gosub.html
Shall I add a link?
I don't know about dynamic form field behaviour without using a gosub. Shall I look through the flow cloud for a more complete picture of uiSchema at this point?

**Comment by lukestanley** [comment]
Linking to that block is a good idea but also yes definitely you should use the general term for it the people are most likely to be familiar with in the first occurrence. I would say at least point out that it works with multiple blocks just like a flow, make sure it doesn't sound like they have to use just one kind of block with it. @CodeKrakken

**Comment by lukestanley** [comment]
Ok I see the basic changes we discussed but what about explaing how it works with data being passed to the embed flows, and other kinds of blocks it supports? @CodeKrakken

**Comment by CodeKrakken** [comment]
> Ok I see the basic changes we discussed but what about explaing how it works with data being passed to the embed flows, and other kinds of blocks it supports? It's still mainly about the gosub blocks (but with more clear language). @CodeKrakken

@Luke I've added a paragraph on how data is passed to the nested flow, but I don't know anything about support for other blocks. Shall I find and study some examples in the Flow Cloud?

**Comment by CodeKrakken** [comment]
> > Ok I see the basic changes we discussed but what about explaing how it works with data being passed to the embed flows, and other kinds of blocks it supports? It's still mainly about the gosub blocks (but with more clear language). @CodeKrakken
> 
> @lukestanley  I've added a paragraph on how data is passed to the nested flow, but I don't know anything about support for other blocks. Shall I find and study some examples in the Flow Cloud?

**Comment by lukestanley** [comment]
> > Ok I see the basic changes we discussed but what about explaining how it works with data being passed to the embed flows, and other kinds of blocks it supports? It's still mainly about the gosub blocks (but with more clear language). @CodeKrakken
 > 
 > @lukestanley  I've added a paragraph on how data is passed to the nested flow, but I don't know anything about support for other blocks. Shall I find and study some examples in the Flow Cloud?
@CodeKrakken 
1. It's too focused on gosub so  mentioning one or two other blocks would help. Think of some that might make sense?
2. It would help to explain how the data flows to the uiSchema. Currently it's too focused on embedded flows, e.g: `uiSchema enables dynamic fields through the insertion or "nesting" of flows into other flows.` makes it sound like this is the only way.

**Comment by CodeKrakken** [comment]
@lukestanley Happy to expand on what's there but I've only used uiSchema once, so that's basically all I can write for now. I'll look through the Flow Cloud for some more applications of it.

**Comment by lukestanley** [comment]
> @lukestanley Happy to expand on what's there but I've only used uiSchema once, so that's basically all I can write for now. I'll look through the Flow Cloud for some more applications of it.

It talks about flows in other flows but it neglects that blocks can contain other blocks without flows, this is not conveyed by what you've written. @CodeKrakken

File: docs/workflow/blocks/form.rst [hunk 1]
```diff
@@ -157,4 +157,190 @@ In the example below, a mapping has a default value, which is saved using the co
...
+  ]
+
+Password Fields
+---------------
+
+You can use uiSchema to designate a password field. 
+The below example will render a form that displays only * for each character entered in the "password" field.
+
+You can see this working `here
+<https://app.kendra.io/bandsintown/connect>`_.
```
Comment (gsambrotta) [inline docs/workflow/blocks/form.rst:169]: :clap: :clap: :clap:  I love that you add an example flow! 
Which also means we have to keep the flow update and running

File: docs/workflow/blocks/form.rst [hunk 2]
```diff
@@ -157,4 +157,190 @@ In the example below, a mapping has a default value, which is saved using the co
...
+Nested flows
+------------
+
+If you want to vary a field according to user input, you can achieve this with a nested flow.
+
+Any flow can be nested in any other flow. The nested flow has access to the main flow's data object, 
+context and state - it can use any data stored here. 
+The nested flow's output is then passed to the main flow's data object, just like the output of a conventional task.
+
+As with the main flow, a nested flow can be edited directly with Kendraio App if opened from the Flow Cloud.
```
Comment (gsambrotta) [inline docs/workflow/blocks/form.rst:257]: This is really great stuff.
Unrelated from this PR but @lukestanley I think we need some text along these lines and style as an introduction in the documentation.

File: docs/workflow/blocks/form.rst [hunk 3]
```diff
@@ -157,4 +157,190 @@ In the example below, a mapping has a default value, which is saved using the co
...
+context and state - it can use any data stored here. 
+The nested flow's output is then passed to the main flow's data object, just like the output of a conventional task.
+
+As with the main flow, a nested flow can be edited directly with Kendraio App if opened from the Flow Cloud. 
+Any saved changes will be reflected immediately when the main flow is refreshed.
+
+This example flow allows the user to search and select from a menu based on returned data. 
+The Venue Name field expects a single value and the Lineup field can handle several values.
+
+The nested flow is denoted by the property ``"type": "gosub"``. You can read more about gosubs :doc:`here <gosub>`.
```
Comment (gsambrotta) [inline docs/workflow/blocks/form.rst:263]: I am not sure, I understood correctly. Does this mean that a nested flow, is a flow that exists already and is "plugged in" in another flow? So I don't find the flow configuration of the nested flow, in the main flow, right?

**Review by gsambrotta — COMMENTED** [review]
Great stuff! It will help, for sure. Just some minors comments

File: docs/workflow/blocks/form.rst (hunk 1 above)
Comment (CodeKrakken) [reply docs/workflow/blocks/form.rst]: @gsambrotta I changed it to a share link, which should mitigate against at least some future breaking changes.

File: docs/workflow/blocks/form.rst [hunk 4]
```diff
@@ -157,4 +157,190 @@ In the example below, a mapping has a default value, which is saved using the co
...
+            "ui:widget": "password"
+        }
+    }
+  }
+
+
+Nested tasks
+------------
+
+You can insert or "nest" another task within a form through the use of uiSchema.
```
[the remark this answers is not in the document]
Comment (CodeKrakken) [reply docs/workflow/blocks/form.rst:202]: @gsambrotta We discussed the name and started calling them tasks when writing articles for the public. Blocks makes more sense to me but I don't know if that's due to the job! I think we should discuss this at the meeting and get the viewpoint of the non tech folks.

File: docs/workflow/blocks/form.rst [hunk 5]
```diff
@@ -157,4 +157,190 @@ In the example below, a mapping has a default value, which is saved using the co
...
+                ]
+            }
+        }
+    }
+}
+
+Nested flows
+------------
+
+If you want to vary a field according to user input, you can achieve this with a nested flow.
```
[the remark this answers is not in the document]
Comment (CodeKrakken) [reply docs/workflow/blocks/form.rst:251]: @gsambrotta Much better, thank you.

File: docs/workflow/blocks/form.rst (hunk 3 above)
Comment (CodeKrakken) [reply docs/workflow/blocks/form.rst:263]: @gsambrotta That's right - it exists independently of the main flow and is edited separately. I'll have a think about how to make this clearer. If you can think of a better way to explain it, please let me know!

Comment (gsambrotta) [reply docs/workflow/blocks/form.rst:264]: Ah OK, all clear. Thank you, also, for making the text more clear.

**Comment by dahacouk** [comment]
Hi All. if you are writing _Flow_ as shorthand for _Kendraio Flow_ then please make sure that the word _Flow_ has a capital _F_, as it is not referring to the _flow of something_ but to a particular concept of a _Flow_ that is special to _Kendraio_. Please do this **_everywhere_**, even in comments and especially in documents. Cheers. cc @lukestanley @gsambrotta @CodeKrakken

File: docs/workflow/blocks/form.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (gsambrotta) [reply docs/workflow/blocks/form.rst:202]: @CodeKrakken I remember we agreed on calling it `block` in the meeting. If you recall it the same, can you please update it?

**Comment by gsambrotta** [comment]
> > @lukestanley Happy to expand on what's there but I've only used uiSchema once, so that's basically all I can write for now. I'll look through the Flow Cloud for some more applications of it.
> 
> It talks about flows in other flows but it neglects that blocks can contain other blocks without flows, this is not conveyed by what you've written. @CodeKrakken

I agree @lukestanley that this documentation is not completed but as we did not use uiSchema much yet, I think it is worthed to start merging this and later add more information as we discover them.

**Comment by gsambrotta** [comment]
@CodeKrakken Can you also please add the Supported Property as you wrote them in the [Google Doc](https://docs.google.com/document/d/1VHIBcXeoKyC8co8Gou7FoXTptydJw2zhEHBu7EnTMEM/edit)?

File: docs/workflow/blocks/form.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (CodeKrakken) [reply docs/workflow/blocks/form.rst]: @gsambrotta So we did - that's done.

**Comment by CodeKrakken** [comment]
@gsambrotta Just rewriting some of these definitions.

**Comment by gsambrotta** [comment]
Awesome, thanks :sparkles:  
Looks good enough to me to be merged. We will improve it and adjust it on the way

**Review by gsambrotta — APPROVED** [review]

**Pull request merged**